### DISCUSSION TITLE
Setting DS3231 programmatically

### DISCUSSION THREAD

**Discussion by PhilTIlson** [micropython/micropython#9876] [Using MicroPython] [answered]
I have a problem with setting a DS3231 RTC clock from within the program.

I am using the ds3231_i2c library and everything is working fine within the program.  Initially I set the clock to the correct time using the recommended code, which is used once then commented out:

`current_time = b'\x30\x52\x12\x03\x03\x11\x22' # sec min hour week day mon year`
`time.sleep(0.1)`
`ds.set_time(current_time)`

Again, this works fine.

However, there may be a situation when the RTC loses the correct time and the user needs to reset it within the program.  My problem is trying to recreate the `current_time = b'\x30\x52\x12\x03\x03\x11\x22'` line.  I'm not very familiar with byte strings and whatever I try to do to create this line fails!  The backslashes seem to be interpreted as escape characters when I try to build the string.

I can easily obtain the necessary settings in the form ['22', '11', '7', '17', '3', '0'] but how do I then convert this into the required form?

I've been struggling over this for a day without success, so I'd be grateful for any advice.

**Comment by davefes** [comment]
Maybe, look at ubinascii

**Comment by scruss** [comment, marked as answer]
Is it this library you're using? [balance19/micropython_DS3231: A library to run the DS3231 RTC on Raspberry Pi Pico](https://github.com/balance19/micropython_DS3231)

That's a really odd way to go about setting the clock. Looks like the library is expecting binary coded decimal.

A couple of routines like this might help:

```py3
def bcd(n):
    n = n % 100
    return 16 * (n // 10) + n % 10

def date_to_bytes(sec, minutes, hour, week, day, month, year):
    # week is week of the year, seemingly
    return bytes(
        [bcd(sec), bcd(minutes), bcd(hour), bcd(week),
         bcd(day), bcd(month), bcd(year)]
    )

# example 2022-11-07 21:25:00 (in week 45)
print(date_to_bytes(0, 25, 21, 45, 7, 11, 22))
```
This returns `b'\x00%!E\x07\x11"'`, which is a slightly more compact way of saying the expected `b'\x00\x25\x21\x45\x07\x11\x22'`.

Alternatively, you could use a different library. Peter Hinch's [ds3231_port.py](https://github.com/peterhinch/micropython-samples/blob/master/DS3231/ds3231_port.py) works well for me, with library docs here: [The DS3231 real time clock chip](https://github.com/peterhinch/micropython-samples/blob/master/DS3231/README.md)

**Reply by PhilTIlson** [reply]
Many thanks for both prompt replies.  **scruss** - that works perfectly!  Thank you.

**Comment by peterhinch** [comment]
See also [this module](https://github.com/peterhinch/micropython-samples/blob/master/DS3231/README.md).

**Reply by scruss** [reply]
I mentioned that one, Peter. It's the one I use.

**Reply by peterhinch** [reply]
Oops! So you did.
